### ISSUE TITLE
High session memory usage caused by SFTP outgoing data buffering

### ISSUE THREAD

**Issue by pepi55** [proftpd/proftpd#1678]
### What I Did

[Using Python's paramiko library](https://pastebin.com/CDNKt1h5), download a file >9GB.
When running this script on the host (downloading from localhost) it causes RAM usage to steadily increase to ridiculously high amounts and 100% cpu spikes on several cores.
In addition, the download takes about 2hrs while the regular sftp client takes about 5-10mins to complete. Sometimes the downloads complete after 2hrs and the file is corrupted/incomplete when using paramiko.

The same script was also tested using sftpgo (default config) and it worked as expected (fast download and complete files).

paramiko version:
paramiko          3.2.0

### What I Expected/Wanted

Download speeds matching unix sftp client, correct download sizes.

### ProFTPD Version and Configuration

```
Compile-time Settings:
  Version: 1.3.7e (maint)
  Platform: LINUX [Linux 5.10.179-166.674.amzn2.x86_64 x86_64]
  Built: Tue May 30 2023 10:43:54 UTC
  Built With:
    configure  '--enable-openssl' '--with-modules=mod_sftp' '--enable-dso' '--with-shared=mod_tls:mod_auth_file' '--disable-auth-pam' '--enable-ctrls' '--enable-ltdl-convenience' '--cache-file=/dev/null' '--srcdir=./' '--enable-devel=nodaemon:nofork:stacktrace'

  CFLAGS: -g2  -Wall -fno-omit-frame-pointer -fno-strict-aliasing -Werror=implicit-function-declaration -Wdangling-else -Wextra -Werror=implicit-function-declaration -Winit-self -Wno-missing-field-initializers -Wno-unused-parameter -Wnull-dereference -fdelete-null-pointer-checks -DPR_DEVEL_NO_FORK -DPR_DEVEL_NO_DAEMON -g3 -O0 -Wcast-align -Wchar-subscripts -Winline -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -Wpointer-arith -Wshadow -Wundef -Wcomment -Wdangling-else -Wdeclaration-after-statement -Wduplicated-branches -Wfloat-equal -Wformat -Wimplicit-function-declaration -Wmaybe-uninitialized -Wmissing-braces -Wparentheses -Wpointer-to-int-cast -Wreturn-local-addr -Wstack-protector -Wstrict-overflow -Wswitch -Wunreachable-code -Wunused-variable -fstack-protector -fstack-protector-all -fno-common
  LDFLAGS: -L$(top_srcdir)/lib -L$(top_builddir)/lib  -rdynamic 
  LIBS:  -lssl -lcrypto -lsodium -lcap  -lcrypto -lsupp -lattr -lnsl -lresolv -lresolv -lcrypt -ldl  

  Files:
    Configuration File:
      /usr/local/etc/proftpd.conf
    Pid File:
      /usr/local/var/proftpd.pid
    Scoreboard File:
      /usr/local/var/proftpd.scoreboard
    Header Directory:
      /usr/local/include/proftpd
    Shared Module Directory:
      /usr/local/libexec

  Info:
    + Max supported UID: 4294967295
    + Max supported GID: 4294967295

  Features:
    - Autoshadow support
    + Controls support
    + curses support
    + Developer support
    + DSO support
    + IPv6 support
    + Largefile support
    - Lastlog support
    - Memcache support
    + ncurses support
    - NLS support
    + OpenSSL support (OpenSSL 1.0.2k  26 Jan 2017, FIPS enabled)
    - PCRE support
    - POSIX ACL support
    - Redis support
    + Sendfile support
    + Shadow file support
    + Sodium support
    + Trace support
    + xattr support

  Tunable Options:
    PR_TUNABLE_BUFFER_SIZE = 1024
    PR_TUNABLE_DEFAULT_RCVBUFSZ = 8192
    PR_TUNABLE_DEFAULT_SNDBUFSZ = 8192
    PR_TUNABLE_ENV_MAX = 2048
    PR_TUNABLE_GLOBBING_MAX_MATCHES = 100000
    PR_TUNABLE_GLOBBING_MAX_RECURSION = 8
    PR_TUNABLE_HASH_TABLE_SIZE = 40
    PR_TUNABLE_LOGIN_MAX = 256
    PR_TUNABLE_NEW_POOL_SIZE = 512
    PR_TUNABLE_PATH_MAX = 4096
    PR_TUNABLE_SCOREBOARD_BUFFER_SIZE = 80
    PR_TUNABLE_SCOREBOARD_SCRUB_TIMER = 30
    PR_TUNABLE_SELECT_TIMEOUT = 30
    PR_TUNABLE_TIMEOUTIDENT = 10
    PR_TUNABLE_TIMEOUTIDLE = 600
    PR_TUNABLE_TIMEOUTLINGER = 10
    PR_TUNABLE_TIMEOUTLOGIN = 300
    PR_TUNABLE_TIMEOUTNOXFER = 300
    PR_TUNABLE_TIMEOUTSTALLED = 3600
    PR_TUNABLE_XFER_SCOREBOARD_UPDATES = 10

```

```
# This is a basic ProFTPD configuration file (rename it to
# 'proftpd.conf' for actual use.  It establishes a single server
# and a single anonymous login.  It assumes that you have a user/group
# "nobody" and "ftp" for normal operation and anon.

#ModulePath %{env:PR_MOD_PATH}

#Include %{env:PR_CONF_PATH}/*.load

#Include %{env:PR_CONF_PATH}/modules/mod_clamav.load
#Include %{env:PR_CONF_PATH}/modules/mod_ban.load
#Include %{env:PR_CONF_PATH}/modules/mod_event.load

ServerName "ProFTPD Not Default Installation"
ServerType standalone
DefaultServer on

MultilineRFC2228 on

Port 0

# Don't use IPv6 support by default.
UseIPv6 off

PersistentPasswd off

#LogFormat auth "%v [%P] %h %t \"%r\" %s"
#LogFormat awstats "%t %h %u %m %f %s %b"
LogFormat auth "{\"remoteHost\":\"%h\",\"time\":\"%{%Y %b %d %H:%M:%S}t\",\"PID\":\"%P\",\"remoteUsername\":\"%l\",\"userID\":\"%u\",\"cmd\":\"%r\",\"status\":\"%s\",\"bytes\":\"%b\"}"
LogFormat awsstats "{\"remoteHost\":\"%h\",\"time\":\"%{%Y %b %d %H:%M:%S}t\",\"PID\":\"%P\",\"remoteUsername\":\"%l\",\"userID\":\"%u\",\"cmd\":\"%r\",\"status\":\"%s\",\"bytes\":\"%b\"}"
LogFormat cloudwatch "{\"remoteHost\":\"%h\",\"time\":\"%{%Y %b %d %H:%M:%S}t\",\"PID\":\"%P\",\"remoteUsername\":\"%l\",\"userID\":\"%u\",\"cmd\":\"%r\",\"status\":\"%s\",\"bytes\":\"%b\"}"
ExtendedLog /var/log/proftpd/proftpd-json.log ALL cloudwatch
SystemLog /var/log/proftpd/proftpd.log

SocketBindTight on
Trace facl:20
Trace DEFAULT:10
#TraceLog /var/log/trace.log
UseReverseDNS off
MaxInstances 400

<Global>
	ExtendedLog /var/log/proftpd/proftpd-json.log ALL cloudwatch

	# Disable .ftpaccess parsing
	AllowOverride off
	AllowOverwrite on
	AllowStoreRestart on
	CapabilitiesEngine on
	DebugLevel 3
	DefaultRoot ~
	DenyFilter \*.*/
	Group proftpd
	ListOptions -l
	RootLogin on
	RootRevoke off
	TimeoutIdle 1200
	TimeoutNoTransfer 86400
	TimeoutStalled 600
	TimesGMT off
	Umask 0
	UseFtpUsers off
	UseSendFile off
	User proftpd
	WtmpLog off
	FSOptions IgnoreExtendedAttributes

	# Deny root by default
	<Limit LOGIN>
		Order Deny,Allow
		DenyUser root
		AllowAll
	</Limit>
</Global>

<IfModule mod_facl.c>
	<Global>
		<Directory />
			HideNoAccess on

			<Limit LIST>
				IgnoreHidden on
			</Limit>
		</Directory>
	</Global>
</IfModule>

<IfModule mod_quotatab.c>
	QuotaEngine off
</IfModule>

<IfModule mod_ratio.c>
	Ratios off
</IfModule>

<IfModule mod_delay.c>
	DelayEngine off
</IfModule>

<IfModule mod_ctrls.c>
	ControlsEngine     off
	ControlsMaxClients 2
	ControlsLog        /var/log/proftpd/controls.log cloudwatch
	ControlsInterval   5
	ControlsSocket     /var/run/proftpd/proftpd.sock
</IfModule>

<IfModule mod_ctrls_admin.c>
	AdminControlsEngine off
</IfModule>

<IfModule mod_sftp.c>
	# Uncomment to run outside of docker
	<VirtualHost localhost %{env:HOSTNAME}>
		#SocketOptions rcvbuf 5000000 sndbuf 5000000
		SFTPEngine on
		#ACLRefreshTime 60
		MaxClientsPerUser 10

		#ExtendedLog /var/log/proftpd/access.log WRITE,READ auth
		ExtendedLog /var/log/proftpd/auth.log AUTH auth
		MaxLoginAttempts 6

		SFTPCiphers aes128-cbc aes128-ctr aes192-cbc aes192-ctr\
			aes256-cbc aes256-ctr arcfour128 arcfour256 blowfish-cbc cast128-cbc

		SFTPClientMatch ".*" channelWindowSize 1GB
		SFTPClientMatch ".*" sftpProtocolVersion 3
		SFTPDHParamFile /etc/dhparams.pem

		Port 2222
		SFTPLog /var/log/proftpd/sftp.log

		# Configure keys
		SFTPHostKey /etc/ssh/ssh_host_rsa_key
		SFTPRekey none

		# Enable compression
		SFTPCompression delayed
		TransferLog /var/log/proftpd/transfer.log
	</VirtualHost>
</IfModule>

#Include %{env:PR_CONF_PATH}/modules/mod_clamav.conf
#Include %{env:PR_CONF_PATH}/modules/mod_event.conf
#Include %{env:PR_CONF_PATH}/modules/mod_ban.conf
```

**Comment by pepi55** [comment]
setting prefetch = False does seem to no longer DOS the server but it is still unreasonably slow compared to paramiko+other servers/other clients+proftpd

**Comment by Castaglia** [comment]
Am I correct in understanding that this ticket is asking about the cause of the unexpectedly high server-side memory usage, seen when using Paramiko to download large files?

**Comment by pepi55** [comment]
I'm not sure why it is happening, no. I'm also not sure if its a config thing I'm missing or if it is a bug. I tried adjusting paramiko to see if its something with the code but as stated before, it works on other servers.

**Comment by Castaglia** [comment]
I currently suspect this is related to https://github.com/paramiko/paramiko/issues/1587

The `prefetch` feature in Paramiko creates a client-side background thread which sends a series of SFTP requests to read "chunks" of the remote file into (client-side) local memory, so that reads from the local memory cache are "fast", to write into the local filesystem.

And this makes sense -- up to a point.  The problem is that SSH, as protocol, has a notion of windowing, just like TCP windows.  (Why?  SSH allows for multiplexing multiple different channels over the same TCP connection, which means it wants better flow control, per channel, over just what TCP provides.)  This Paramiko `prefetch` feature leads to cases where the client sends SFTP `READ` requests for more data than the SSH channel allows.

When this happens, a server has 2 choices:
* Stop reading SFTP requests until the window opens again (which is what OpenSSH does)
* Buffer up the generated SFTP responses until the window opens again (which is what `mod_sftp` does)

Either way, client-side, this leads to the impression of "hanging", because either way, the server is *not* sending the requested data (because the client, Paramiko, is requesting more data than can actually be provided).

The recently-added Paramiko `max_concurrent_prefetch_requests` allows for setting a client-side bound on the number of these `prefetch` requests, which can help, yes.  However, this still points out that there is the _server-side_ problem, where clients request more data than can be sent to them.

I'm still not sure yet of a "good" solution to this particular problem.  As the multiple Paramiko StackOverflow (and other) posts show, this is indeed a likely scenario.

I suspect that whatever mitigation I devise will not be great from a user experience perspective, because inevitably it will lead to "hanging", or an "aborted" download.  Not fun.

**Comment by pepi55** [comment]
So if I understand correctly, the current implementation of mod_sftp allows for transfers _without_ hanging at the cost of increased server load?

However, if im misunderstanding something and the transfer hangs either way, I think it would be a good idea to at least prevent the steadily increasing memory usage, right? This feels like a paramiko issue and therefore I think paramiko is responsible for creating a smooth user experience. As mentioned before, using paramiko is practically the only case in which the server gobbles up all the RAM like that.

**Comment by Castaglia** [comment]
There's no current option to prevent this buffering, no.  Any mitigating measure that I can currently think of will result in the perception of "hanging downloads" or "stuck" downloads for Paramiko users (or any other client that ends up with similar behavior).  I'm trying to determine which is the best, of the worst possibility changes, to try to add to `mod_sftp`.

**Comment by pepi55** [comment]
Well if its any help: i tried using the script with openSSH and while the tragic download speed is consistent, the resource usage on the server side is what you would expect from an sftp transfer (<100mb of RAM usage).

the terrible client side experience is something I would consider a problem of paramiko and while sftpGO probably does some fancy behind the scenes stuff to compensate for faulty clients, all I need of mod_sftp is to stay within reasonable system resource usage.

I appreciate you taking the time to look into this problem

**Comment by Castaglia** [comment]
@pepi55 I've had success with https://github.com/proftpd/proftpd/pull/1699 locally.  With those changes, the ProFTPD session memory usage remains constantly low (relatively), and no longer balloons up.  The Paramiko transfer latency increases, understandably, but will still eventually succeed.

Care to try it out in your setup, make sure that it works for you?

The solution I went for is to buffer up the incoming SFTP requests (the "preflight" READs that Paramiko sends for every chunk of the file to be downloaded), if the channel window closes.  We thus do not even process those READ requests (and try to send the requested chunk), unless we have room in that window.  We will process the "window adjustment" requests that Paramiko sends, to open the window once more.  And upon receiving such a "window adjustment", we proceed to drain/process those pending READ requests.  All of this should be invisible to the client, at the cost of increased latency (due to waiting for the client's "window adjustment" requests).

**Comment by pepi55** [comment]
Awesome, I will give it a try coming monday and let you know how it goes!

**Comment by pepi55** [comment]
This solution seems to resolve the issue I have been having, however I get a stacktrace after a while:
127.0.0.1 (10.178.244.211[10.178.244.211]): -----BEGIN STACK TRACE-----
127.0.0.1 (10.178.244.211[10.178.244.211]): -----END STACK TRACE-----
127.0.0.1 (10.178.244.211[10.178.244.211]): ProFTPD terminating (signal 11)
127.0.0.1 (10.178.244.211[10.178.244.211]): SSH2 session closed.
127.0.0.1 (10.178.244.211[10.178.244.211]): ProFTPD terminating (signal 11)
127.0.0.1 (10.178.244.211[10.178.244.211]): SSH2 session closed.

not sure what that is about... any clue where I should be looking for more info on this stacktrace?

**Comment by Castaglia** [comment]
OK, thanks for the report.  At least we're making some progress toward better behavior!

When you say you get a stacktrace "after a while", can you elaborate?  That is, can you say whether it happens toward the end of transferring a given file, or maybe it happens at the end of the session, after all files have been transferred?  Maybe it's only after the 3rd or 4th file?  The _timing_ of when the stacktrace occurs can give me some clues of how to reproduce (and then fix) that behavior locally.  Thanks!

**Comment by pepi55** [comment]
the openSSH client has no issues downloading (100%   10GB  11.1MB/s   15:19)
WinSCP also has no issues (100%   10GB  11.0MB/s   15:21)

so that leaves paramiko again.

Smaller files download successfully:
```
$ ./DOSAttack.py
>>> downloading mb13.bin at: 2023/8/16 12:42:08
completed at: 2023/8/16 12:42:09
download time: 0:00:01.228280
>>> downloading MB138.zip at: 2023/8/16 12:42:09
completed at: 2023/8/16 12:42:16
download time: 0:00:07.197818
>>> downloading oneGB.bin at: 2023/8/16 12:42:16
completed at: 2023/8/16 12:46:23
download time: 0:04:06.922362
>>> downloading 5gb.bin at: 2023/8/16 15:09:37
completed at: 2023/8/16 17:02:10
download time: 1:52:33.077648
```

however, with bigger files, weird things start happening:
```
>>> downloading GB10.bin at: 2023/8/16 17:02:10
Socket exception: An existing connection was forcibly closed by the remote host (10054)
download failed at: 2023/8/16 17:23:53
elapsed time: 0:21:42.489637
Traceback (most recent call last):
  File "C:\Users\user\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\paramiko\sftp_client.py", line 852, in _read_response
    t, data = self._read_packet()
  File "C:\Users\user\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\paramiko\sftp.py", line 218, in _read_packet
    data = self._read_all(size)
  File "C:\Users\user\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\paramiko\sftp.py", line 199, in _read_all
    raise EOFError()
EOFError
```

```
127.0.0.1 (10.207.14.90[10.207.14.90]): -----BEGIN STACK TRACE-----
127.0.0.1 (10.207.14.90[10.207.14.90]): -----END STACK TRACE-----
127.0.0.1 (10.207.14.90[10.207.14.90]): ProFTPD terminating (signal 11)
127.0.0.1 (10.207.14.90[10.207.14.90]): SSH2 session closed.
127.0.0.1 (10.207.14.90[10.207.14.90]): ProFTPD terminating (signal 11)
127.0.0.1 (10.207.14.90[10.207.14.90]): SSH2 session closed.
```

I tried with another big file too:

```
>>> downloading tenGB.bin at: 2023/8/16 18:28:33
Socket exception: An existing connection was forcibly closed by the remote host (10054)
download failed at: 2023/8/16 18:49:30
elapsed time: 0:20:56.639734
tenGB.bin failed, trying next file...
```

seems like it fails at around the 20min mark but only if the file is sufficiently large. I have never seen such an issue before.

**Comment by Castaglia** [comment]
I suspect that neither OpenSSH nor WinSCP have this "preflight" behavior that Paramiko does; those clients are _much_ better behaved, protocol-wise -- and thus less likely to encounter this particular issue (or code path).

No client should Ideally be able to trigger a segfault in the server, so I definitely want to track this down.  I'm wondering if it's related to the size of the file (although I successfully tested this patch locally using a 10GB file), or the _time_ it takes to do the transfer (maybe longer is more likely to encounter this)...or both.  I'll see what other crazy combinations I can do locally as well.

Are you still using these particular `mod_sftp` settings?
```
		SFTPClientMatch ".*" channelWindowSize 1GB
		SFTPClientMatch ".*" sftpProtocolVersion 3

		SFTPRekey none

		# Enable compression
		SFTPCompression delayed
```
And is Paramiko using compression for its transfers?

**Comment by pepi55** [comment]
yes, except: `SFTPClientMatch ".*WinSCP.*" sftpProtocolVersion 3` and I haven't enabled compression for paramiko.
im going to try again with `SFTPClientMatch ".*" sftpProtocolVersion 3` and with compression.

a note on the time aspect: the 5GB file transferred successfully in 1 hour and 52 minutes. I think it is related to a combination of size and maybe throughput or something like that. the download speed was about 11MB/s since i connect remotely and not locally. I could run a local download as well but that does not really reflect the use case.

**Comment by pepi55** [comment]
```
$ ./DOSAttack.py
>>> downloading tenGB.bin at: 2023/8/16 20:07:16
Socket exception: An existing connection was forcibly closed by the remote host (10054)
download failed at: 2023/8/16 20:25:31
elapsed time: 0:18:14.538087
tenGB.bin failed, trying next file...
```

it also seems to consistently fail at between 
![image](https://github.com/proftpd/proftpd/assets/5328981/0551bd29-1dee-44b1-b7e9-d29392a8c20c)
and ~52,000KB

**Comment by Castaglia** [comment]
Here's how I usually try to track down where a segfault is ocurring in ProFTPD.

When building from source, I only use static modules, not shared modules.  In addition, I make sure to enable the code that makes stacktraces easier to diagnose.  Thus:
```sh
cd proftpd
make clean
./configure --with-modules=mod_sftp --enable-devel=stacktrace
make
make install
```
Then when running ProFTPD, I run it under a debugger, using the `-X` command-line flag to tell ProFTPD _not_ to fork a separate process to handle a connection.  This is great for testing and debugging, but not something to be used for servers where you _must_ handle multiple connections:
```
$ gdb /usr/local/sbin/proftpd
(gdb) run -n -X
...
```
Then, separately, you run your test (in your case, @pepi55 , it'd be `./DOSAttack.py`).  Eventually, ProFTPD _should_ segfault.  You'll see a message in the debugger about where the process halted.

When this happens, type:
```
(gdb) bt
```
The `bt` is shorthand for "backtrace".  It should print out all of the function calls (including file names _and_ line numbers!) that lead to the problem.

If you're able to do this, and you get the screenshots/text from `gdb`, that would be very useful!

**Comment by pepi55** [comment]
Okay, i'll give this a try and let you know how it goes.

**Comment by pepi55** [comment]
this is the backtrace I got:
```
(gdb) bt
#0  0x00000000004216b4 in make_sub_pool (p=0x7dc89d0) at pool.c:502
#1  0x00000000004d4f2b in log_event (cmd=0x7dc89a8, lf=0x8701d0) at mod_log.c:752
#2  0x00000000004d5245 in log_any (cmd=0x7dc89a8) at mod_log.c:820
#3  0x00000000004d56d2 in log_xfer_stalled_ev (event_data=0x0, user_data=0x0) at mod_log.c:957
#4  0x0000000000472bb6 in pr_event_generate (event=0x594215 "core.timeout-stalled", event_data=0x0) at event.c:371
#5  0x000000000052d49c in fxp_timeout_stalled_cb (p1=600, p2=4, p3=600, data=0x0) at fxp.c:360
#6  0x000000000041f369 in process_timers (elapsed=1) at timers.c:140
#7  0x000000000041f791 in handle_alarm () at timers.c:287
#8  0x000000000041ff08 in pr_alarms_unblock () at timers.c:518
#9  0x000000000041f927 in pr_timer_reset (timerno=2, mod=0xffffffff) at timers.c:336
#10 0x00000000004fa584 in sftp_ssh2_packet_read (sockfd=0, pkt=0x8731f8) at packet.c:1232
#11 0x00000000004fc0d7 in sftp_ssh2_packet_process (p=0x86f9b0) at packet.c:1996
#12 0x00000000004ee852 in sftp_cmd_loop (s=0x816b28, conn=0x86c8f8) at mod_sftp.c:331
#13 0x000000000041d38f in fork_server (fd=8, l=0x86c498, no_fork=1 '\001') at main.c:1519
#14 0x000000000041dbc2 in daemon_loop () at main.c:1764
#15 0x000000000041e06f in standalone_main () at main.c:1965
#16 0x000000000041f075 in main (argc=5, argv=0x7fffffffdc18, envp=0x7fffffffdc48) at main.c:2790
(gdb)
```
again on the ~20th minute mark.

**Comment by Castaglia** [comment]
Thanks!  That stacktrace does help a bit.  In particular, I see:
```
#4  0x0000000000472bb6 in pr_event_generate (event=0x594215 "core.timeout-stalled", event_data=0x0)
```
This indicates that at this time, your [`TimeoutStalled`](http://www.proftpd.org/docs/modules/mod_xfer.html#TimeoutStalled) timer triggered.  I'll see if I can configure my local `TimeoutStalled` such it'll trigger during my local Paramiko setup, and see if that helps me reproduce this.

**Comment by Castaglia** [comment]
I finally made more progress on this issue; I'm able to reproduce the stacktrace issue locally.  I'm not sure why yet, but it appears related to the `LogFormat`/`ExtendedLog` configuration being used, in conjunction with this Paramiko behavior.

**Comment by Castaglia** [comment]
@pepi55 I've updated the patch with a change that hopefully fixes/address the stacktrace behavior you're seeing.  Could you try out the latest patch, and see how it behaves for you?

**Comment by Castaglia** [comment]
As this patch is working for my local tests, I'm going to merge it.  We can fine tune and polish any issues as they come up.  Thanks!

**Comment by Castaglia** [comment]
PR has been merged to master, and backported to the 1.3.8 branch.  I'm going to close this ticket; we can re-open it if necessary.

**Comment by pepi55** [comment]
Hi, sorry for the late response but I was away for 2 weeks. I will check if I get the issue now. Should I just open a new issue if another problem arises or keep the discussion here?

**Comment by Castaglia** [comment]
@pepi55 No worries; I hope you had a good holiday.

If you encounter any issues, it's fine to keep the discussion here; we can re-open this ticket if needed.  Thanks!

**Comment by Malizor** [comment]
I everyone,

Since version 3.3, Paramiko added the `max_concurrent_prefetch_requests` parameter, with an interesting note in the documentation:

> OpenSSH’s sftp internally imposes a limit of 64 concurrent requests, while Paramiko imposes no limit by default; consider setting a limit if a file can be successfully received with sftp but hangs with Paramiko.

I understand this is not the approach ProFTPd took, right?
In this case, is their any value you would recommend for use with ProFTPd?

**Comment by Castaglia** [comment]
> > OpenSSH’s sftp internally imposes a limit of 64 concurrent requests, while Paramiko imposes no limit by default; consider setting a limit if a file can be successfully received with sftp but hangs with Paramiko.
> 
> I understand the OpenSSH approach is not the one ProFTPd took, right? In this case, is their any value you would recommend for use with ProFTPd?

Correct, we took a different approach to handling behavior like Paramiko's (prior to its `max_concurrent_prefetch_requests` addition).  The `mod_sftp` module pays attention to the SSH channel window; when the window closes, any additional incoming requests are buffered, and are drained/processed once the remote client (Paramiko) indicates space in that window once more.  This buffering does mean that memory will be used -- just not as much as before.

The end result is behavior that is similar to OpenSSH's, just done via a different mechanism/implementation.

> Here (Alma 9, ProFTPd 1.3.8b), all the server RAM is still consumed when launching a single download of a big file (ie. size > RAM size) via Paramiko.

The code change for this issue first appeared in ProFTPD 1.3.9rc1, and not in earlier versions.

**Comment by Malizor** [comment]
Thanks for your answer and the explanation.

> The code change for this issue first appeared in ProFTPD 1.3.9rc1, and not in earlier versions.

Well, you stated that the fix was "backported to the 1.3.8 branch" and it is mentioned in the 1.3.8a changelog.
Is it not the full fix? Here I do not see much behavioral difference between 1.3.8b and the 1.3.7 branch.

**Comment by Castaglia** [comment]
> Here I do not see much behavioral difference between 1.3.8b and the 1.3.7 branch.

Can you describe the situation/setup you have there in more detail, and what and how you are observing the behavior?

**Comment by Malizor** [comment]
Thanks a lot for the explanation @Castaglia !

Here is what I observe on my test setup (2GB of RAM, test by downloading a 10GB file with default Paramiko):

| `channelWindowSize` value | Used memory, measured with `htop` |
|--------|--------|
| 16MB and more | ~33% |
| 8MB | ~20% |
| 4MB | ~11% |
| 2MB | ~5% |
| 1MB | ~3% | 

Strangely enough, the transfer speed remained consistent in all cases (~2 minutes ±3 seconds).
So I don't see any drawback?
Same with a Filezilla client, no download speed difference between a 1GB and 1MB `channelWindowSize`.
Actually Filezilla was slower while uploading with a 1GB window size.

Our use case here is for an SFTP server used by third parties which use whatever client they want (and may change at anytime).
So we must have default settings which deal with any ill-behaved client (preventing the server from filling the memory and crashing).
The fact that the main Python library for doing SFTP is not "well behaved" just made us notice the problem.
So I guess we are going with a combination of a small `channelWindowSize` and a limited `MaxClients`.

However I don't think it is great that ProFTPd default settings means that an unexpected amount of memory can be used if the "wrong" client is used. I think the documentation should at least talk about it.
I would also suggest using a way lower default `channelWindowSize`, like 1 or 2 MB, since I did not see any drawback in my testing. But maybe I missed something.

**Comment by Castaglia** [comment]
> However I don't think it is great that ProFTPd default settings means that an unexpected amount of memory can be used if the "wrong" client is used. I think the documentation should at least talk about it. I would also suggest using a way lower default `channelWindowSize`, like 1 or 2 MB, since I did not see any drawback in my testing. But maybe I missed something.

If you want to use a smaller channel window size for any/all clients, you can use _e.g_:
```
SFTPClientMatch .* channelWindowSize 2MB
```

**Comment by halida** [comment]
I set channelWindowSize but it still use all my server memory, not sure what is wrong?

SFTPClientMatch .* channelWindowSize 200KB
